draco/analysis/flagging.py

```diff
@@ -628,49 +627,56 @@ class ThresholdVisWeight(task.SingleTask):
     """
 
     absolute_threshold = config.Property(proptype=float, default=1e-7)
-    relative_threshold = config.Property(proptype=float, default=0.0)
+    relative_threshold = config.Property(proptype=float, default=0.9)
 
-    def process(self, timestream):
-        """Apply threshold to `weight` dataset.
+    def process(self, stream):
+        """Make a baseline-independent mask.
 
         Parameters
         ----------
         timestream : `.core.container` with `weight` attribute
 
         Returns
         -------
-        timestream : same as input timestream
-            The input container with modified weights.
+        out : RFIMask container
+            RFIMask container with mask set
         """
-        from mpi4py import MPI
-
-        timestream.redistribute(["prod", "stack"])
-
-        weight = timestream.weight[:]
-
-        # Average over the frequency and time axes to get a per baseline
-        # average
-        mean_weight = weight.mean(axis=2).mean(axis=0)
+        stream.redistribute("freq")
 
-        # Figure out which entries to keep
-        threshold = np.maximum(
-            self.absolute_threshold, self.relative_threshold * mean_weight
+        # Make the output container depending on what 'stream' is
+        if "ra" in stream.axes:
+            out = containers.SiderealRFIMask(axes_from=stream, attrs_from=stream)
+        elif "time" in stream.axes:
+            out = containers.RFIMask(axes_from=stream, attrs_from=stream)
+        else:
+            raise TypeError(f"Require Timestream or SiderealStream. Got {type(stream)}")
+
+        weight = stream.weight[:]
+        # Average over the baselines (stack) axis
+        mean_baseline = np.mean(weight, axis=1, keepdims=True)
+        # Cut out any values below fixed threhold. Return a np.ndarray
+        threshold = np.where(
+            mean_baseline > self.absolute_threshold, mean_baseline, np.nan
         )
-        keep = weight > threshold[np.newaxis, :, np.newaxis]
-        keep_sum = np.sum(keep)
-        keep_total = np.zeros_like(keep_sum)
-
-        timestream.comm.Allreduce(keep_sum, keep_total, op=MPI.SUM)
-        keep_frac = keep_total / float(np.prod(weight.global_shape))
-
-        self.log.info(
-            "%0.5f%% of data is below the weight threshold"
-            % (100.0 * (1.0 - keep_frac))
+        # Average across the time (ra) axis to get per-frequency thresholds,
+        # ignoring any nans
+        threshold = np.nanmean(threshold, axis=2, keepdims=True)
+        # Create a 2D baseline-independent mask.
+        mask = ~np.squeeze(
+            mean_baseline.local_array > threshold * self.relative_threshold
```

Comment:
I don't think the criterion actually works as written.

Take a single frequency (after baseline averaging) with two values, one a very small value epsilon below the absolute threshold and one epsilon above, i.e. `[t - eps, t + eps]`. You end up with threshold on (L658) becoming `[NaN, t + eps]`, then on L663 after the nanmean becoming `[t + eps]`. A typical value of relative_threshold might be ~0.5, and provided `eps < t * 0.5` *neither* of the times within that one frequency is masked.

I think what you need to do is either generate masks for both relative and absolute thresholds and then combine, or  I think you can just do `mean_baseline.local_array > np.fmax(threshold * self.relative_threshold, self.absolute_threshold)`